### ISSUE TITLE
Layer update feature

### ISSUE THREAD

**Issue by echeipesh** [geotrellis/geotrellis#1187]
Layer update should allow to update an already saved layer with a smaller layer that may intersect values already saved. This needs to be done at minimum on a level thats generic on 'K' and 'V'. It would be a bonus if it's possible to make it generic on the backend, some kind of composition over generic layer reader and writer.

### Update Workflow
- Verify layer `keyClass` and `valueClass` match our instance type
- Get `KeyBounds[K]` of update tiles
- Load `KeyIndex[K]` of existing layer
- Verify `schema` matches existing layer
- Bonus: It would be nice if we wrote `RDDWriter/RDDReader` used for the layer
- Update layer KeyBounds and MetaData
- Write updated KeyBounds MetaData
- Writer tiles using `RDDWriter`

It probably makes most sense to subclass `Writer` into `UpdatingWriter` that has an `def update(id, value): Unit` method.

### Use Case
```scala
val reader = S3LayerReader[SpatialKey, Tile, RasterRDD]("bucket","key")
val writer = S3LayerWriter[SpatialKey, Tile, RasterRDD]("bucket", "key", ZCurveIndexMethod)

val id: LayerId = ???
val updateRDD: RasterRDD[SpatialKey] = ???

// Option 1:
writer.update(id, updateRDD)

// Option 2:
writer.update(id, updateRDD, reader)
```

Of course in order to update we need a reader that will be used to read the tiles, if any, that are already written so they can be merged with new tiles.

One weirdness is that the writer is going to have to disregard it's constructor argument for `keyIndexMethod` and use whatever `KeyIndex` is already in the `attributeStore`.
Maybe that constructor argument should be renamed to `defaultKeyIndexMethod`.

#### Option 1
The writer nearly has everything in scope it needs to create a `S3LayerReader`. Likely `RDDWriter` and `RDDReader` traits need to be merged into one, `RDDFormat`.
One problem with this is that the `attributeStore` on another reader can potentially be stale. This can be solved using shared cache for the `AttributeStore`, something that keys on `bucket` and `prefix` so any instance of `S3AttributeStore` that points to the same location shares a cache.

#### Option 2
Can side-step the option and ask for a reader we need to query. `LayerReader` trait can include `attributeStore` field so it can be used for writing the new metadata, thus updating the cache.

### Notes
The implementation can ask for `Boundable[K]` (LayerWriter already asks for this) type-class which will give the ability to collect `KeyBounds[K]` of the argument and to
merge the them with already saved `KeyBounds[K]`.

It can also ask for `MergeView[V]` to merge tiles of underlying type. This type class is used in `Tiler` (for reference). Not sure if this should go into constructor or update method.

Since MetaData is treated existentially through the `ContainerConstructor` it will have to gain a method to combine them.

### Next Steps

Extend `Pyramid` object so gives a sequence of RDDs that can be used for an update:

```scala
  val updateRDD: RasterRDD[SpatialKey] = ???
  val writer = S3LayerWriter[SpatialKey, Tile, RasterRDD]("bucket", "key", ZCurveIndexMethod)
  
  Pyramid.up(updateRDD, 1).foreach { case (zoom, rdd) =>
    writer.update(LayerId("my-layer", zoom), rdd)
  }
```

**Comment by pomadchin** [comment]
Should update work as a `writeOrUpdate` function? Or it only should be used for an layer update?

**Comment by echeipesh** [comment]
To curate gitter conversation:

Update should only be available for layers that already exist and should follow this use case:

```scala
val reader = S3LayerReader[SpatialKey, Tile, RasterRDD]("bucket","key")
val writer = S3LayerWriter[SpatialKey, Tile, RasterRDD]("bucket", "key", ZCurveIndexMethod)

val update: RasterRDD[SpatialKey] = ??? // got it this from somewhere
val id: LayerId = ??? // Id of layer we are trying to update

// explicit version
val existing = reader.query(id).where(Intersects(update)).toRDD
val merged = existing merge update
writer.update(id, merged)

//sugared version:
writer.mergeUpdate(reader, id) { tilesToUpdate => tilesToUpdate merge update }
```

**Comment by lossyrob** [comment]
This was closed with PR #1192
